### PR TITLE
Add mutate_apply specialization that can retrieve a tag

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Instead of providing an invokable as a template argument to `db::mutate_apply` directly, this PR also allows providing a tag that holds an invokable. The invokable will be retrieved from the DataBox and may be stateful.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/DataStructures/DataBox/DataBox.hpp [hunk 1]
```diff
@@ -1888,35 +1888,76 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
...
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Apply the invokable retrieved from the `InvokableTag` mutating the
+ * `InvokableTag::type::return_tags` and taking as additional arguments the
+ * `InvokableTag::type::argument_tags` and `args`.
+ *
+ * \details
+ * `InvokableTag::type` must have `tmpl::list` type aliases `return_tags` and
+ * `argument_tags`, as well as a static `apply` function. The `apply` function
```
Comment (wthrowe) [inline src/DataStructures/DataBox/DataBox.hpp]: This isn't very useful if it can only call static functions.

Comment (nilsvu) [reply src/DataStructures/DataBox/DataBox.hpp]: That was a mistake in the docs; updated it. This function just forwards to the standard stateful `db::mutate_apply` overload, so a call operator or a static apply function is fine.

**Comment by nilsdeppe** [comment]
Why do we want this? I don't like it. It's overusing the DataBox even more in my opinion and will just give me more things to remove later...

**Comment by nilsvu** [comment]
It allows to add the following to an action list:

```cpp
Actions::MutateApply<computer_tag>
```

where now the `computer_tag::type` can be stateful, e.g. constructed from options. Useful for #1668 for example.

**Comment by nilsdeppe** [comment]
That should be explicitly done by an action. That's why this doesn't exist. It makes understanding DataBox usage even more difficult.

**Comment by nilsvu** [comment]
#1668 currently adds a wrapper around such a stateful `computer_tag` just to fetch the tag. Would be good to make that easier.

**Comment by nilsdeppe** [comment]
Yes, let's not enable this type of thing more than it already is. It encourages even more abuse of the DataBox.

**Comment by kidder** [comment]
I don't think this is the direction we want to go...

**Comment by nilsvu** [comment]
Closing for now then

**Pull request closed without merging**